middleware/src/authentication/authentication.go

```diff
@@ -0,0 +1,95 @@
+package authentication
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"log"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+)
+
+type Claims struct {
+	Username string `json:"username"`
+	jwt.StandardClaims
+}
+
+type JwtAuth struct {
+	jwtKey string
+}
+
+func NewJwtAuth() *JwtAuth {
+	jwtAuth := &JwtAuth{}
+	// generate random string with 32 bytes of entropy and use it as the signing key
+	JwtKey, err := jwtAuth.generateRandomString(32)
+	if err != nil {
+		log.Println("could not get enough entropy to generate key")
+	}
+	jwtAuth.jwtKey = JwtKey
+	return jwtAuth
+}
+
+// GenerateRandomBytes returns securely generated random bytes. It will return an error
+// if the system's secure random number generator fails to function correctly, in which
+// case the caller should not continue. These utility functions are taken from:
+// https://stackoverflow.com/questions/32349807/how-can-i-generate-a-random-int-using-the-crypto-rand-package
+func (jwtAuth *JwtAuth) generateRandomBytes(n int) ([]byte, error) {
+	b := make([]byte, n)
+	_, err := rand.Read(b)
+	// Note that err == nil only if we read len(b) bytes.
+	if err != nil {
+		log.Println("unable to read random bytes during jwt key generation")
+		return nil, err
+	}
+
+	return b, nil
+}
+
+// GenerateRandomString returns a URL-safe, base64 encoded securely generated random string.
+func (jwtAuth *JwtAuth) generateRandomString(s int) (string, error) {
+	b, err := jwtAuth.generateRandomBytes(s)
+	return base64.URLEncoding.EncodeToString(b), err
+}
+
+// GenerateToken generates a jwt token. It is the callers job to ensure that the username has been verified in the database first.
+func (jwtAuth *JwtAuth) GenerateToken(username string) (string, error) {
+	jwtStaticKey := []byte(jwtAuth.jwtKey)
+
+	// TODO: Implement proper handling of the expiration time. This is basically just a dummy value 24 hours in to the future.
```

Comment:
What are your thoughts here regarding:
- When should this be done?
- What would a good expiry time?

Reply:
Pulling @Stadicus and @KasparEtter into this discussion.
NIST recommends an absolute maximum of 30 days for the lifetime of an authentication token. However mobile apps for example never require their users to re-authenticate manually. To enact this they use so called 'refresh tokens' which have a far longer expiry time to get a new authentication token once the authentication token has expired. This refresh token is used once and then invalidated. I don't like this though, in my eyes it just adds a layer of complexity and obscurity. It is possible to do correctly through oauth, but then requires a 3rd party to function properly. It's also a bad idea within the current architecture, since the token is no longer valid if the middleware has to restart (it generates a new jwt key on every restart). 
Since jwt makes it hard to log out users from a session, it seems to be preferable in my eyes to have a fairly low expiry time and then ask the user to re-authenticate with his password once the jwt token expires. I think a day is reasonable in this case, since we have additional security guarantees through the usage of the noise pairing. What are your opinions on this matter?

Reply:
After some offline discussion I'll opt to go with how it is done currently. A user will be forced to re-authenticate after 24h, notifications will still be streamed users even after their token is expired. Eventually there could be some rpc calls that are not authenticated that provide basic functionality (like an electrs proxy)

Reply:
`TODO` can be removed?

Reply:
Done.